lib/usdUfe/ufe/UsdUndoDuplicateCommand.cpp

```diff
@@ -0,0 +1,159 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "UsdUndoDuplicateCommand.h"
+
+#include <usdUfe/base/tokens.h>
+#include <usdUfe/ufe/Global.h>
+#include <usdUfe/ufe/Utils.h>
+#include <usdUfe/undo/UsdUndoBlock.h>
+#include <usdUfe/utils/editRouter.h>
+#include <usdUfe/utils/editRouterContext.h>
+#include <usdUfe/utils/layers.h>
+#include <usdUfe/utils/loadRules.h>
+#include <usdUfe/utils/mergePrims.h>
+
+#include <pxr/base/tf/token.h>
+#include <pxr/usd/sdf/copyUtils.h>
+#include <pxr/usd/usd/prim.h>
+#include <pxr/usd/usd/stage.h>
+
+namespace USDUFE_NS_DEF {
+
+// Ensure that UsdUndoDuplicateCommand is properly setup.
+static_assert(std::is_base_of<Ufe::UndoableCommand, UsdUndoDuplicateCommand>::value);
+static_assert(std::has_virtual_destructor<UsdUndoDuplicateCommand>::value);
+static_assert(!std::is_copy_constructible<UsdUndoDuplicateCommand>::value);
+static_assert(!std::is_copy_assignable<UsdUndoDuplicateCommand>::value);
+static_assert(!std::is_move_constructible<UsdUndoDuplicateCommand>::value);
+static_assert(!std::is_move_assignable<UsdUndoDuplicateCommand>::value);
+
+namespace {
+
+// Local ufePathToPrim that uses input stage instead of getStage function.
+UsdPrim ufePathToPrim(const Ufe::Path& path, const PXR_NS::UsdStageWeakPtr& stage)
+{
+    if (!stage)
+        return {};
+
+    const Ufe::Path            ufePrimPath = stripInstanceIndexFromUfePath(path);
+    const Ufe::Path::Segments& segments = ufePrimPath.getSegments();
+    if (segments.empty())
+        return {};
+
+    // Anonymous layer.
+    if (path.string().at(0) == '/') {
+        return stage->GetPrimAtPath(SdfPath(path.string()).GetPrimPath());
+    }
+
+    // If there is only a single segment in the path, it must point to the
+    // proxy shape, otherwise we would not have retrieved a valid stage.
+    // The second path segment is the USD path.
+    return (segments.size() == 1U)
+        ? stage->GetPseudoRoot()
+        : stage->GetPrimAtPath(SdfPath(segments[1].string()).GetPrimPath());
+}
+
+} // namespace
+
+UsdUndoDuplicateCommand::UsdUndoDuplicateCommand(
+    const UsdSceneItem::Ptr& srcItem,
+    const UsdSceneItem::Ptr& dstParentItem)
+    : _ufeDstPath(dstParentItem->path())
+    , _ufeSrcPath(srcItem->path())
+    , _dstStage(dstParentItem->prim().GetStage())
+    , _srcStage(srcItem->prim().GetStage())
+{
+    auto srcPrim = srcItem->prim();
+    auto newName = uniqueChildName(dstParentItem->prim(), srcPrim.GetName());
+    _usdDstPath = dstParentItem->prim().GetPath().AppendChild(TfToken(newName));
+}
+
+UsdUndoDuplicateCommand::Ptr UsdUndoDuplicateCommand::create(
+    const UsdSceneItem::Ptr& srcItem,
+    const UsdSceneItem::Ptr& dstParentItem)
+{
+    return std::make_shared<UsdUndoDuplicateCommand>(srcItem, dstParentItem);
+}
+UsdSceneItem::Ptr UsdUndoDuplicateCommand::duplicatedItem() const
+{
+    Ufe::Path ufeSrcPath;
```

Comment:
This function assumes that the target did not get rename due to name collision.

Reply:
Doesn't `_usdDstPath` contain the renamed prim?